### DISCUSSION TITLE
Proper usage of changeUnit for solely index creation

### DISCUSSION THREAD

**Discussion by nazeem-soeltan** [flamingock/mongock#47] [Q&A] [answered]
Hi,

I'm fiddling around with using the new `ChangeUnit` concept and creating indices.
So I figured out that indices can not be created within a transaction, as a limitation of Mongo `4.2.x`. (`4.4.x` does support it).

But for now, What I thought is the following. As per my understanding the `@BeforeExecution` and `@RollbackBeforeExecution` are outside of a transaction, thus allowing these kind of operations:

**ChangeUnit**

```java
@ChangeUnit(id = "add-index-on-company-name", order = "2", author = "nazeem.soeltan")
public class AddIndexOnCompanyName {

    @BeforeExecution
    public void beforeExecute(final MongoTemplate mongoTemplate) {
        final var index = new Index()
                .background()
                .named("name_index")
                .on("name", DEFAULT_DIRECTION);

        mongoTemplate.indexOps(Company.class).ensureIndex(index);
    }

    @RollbackBeforeExecution
    public void rollbackBeforeExecute() {
        // omit.
    }

    @Execution
    public void execute(final MongoTemplate mongoTemplate) {
        // omit.
    }

    @RollbackExecution
    public void executeRollback(final MongoTemplate mongoTemplate) {
        // omit.
    }

}
```

However, what I noticed is that this creates two entries in `mongockChangeLog`. The `add-index-on-company-name_before` and for `add-index-on-company-name`, where the latter one is obviously "empty" as no change has actually been done there.
In a way, this creates some 'pollution'. I'm wondering, is the approach that I described above, the correct one (given the limitation of 4.2), or is there some kind of alternative which doesn't create this extra 'empty' entry in the table? 

Thanks in advance.

**Comment by dieppa** [comment, marked as answer]
Hello @nazeem-soeltan ,

First, keep in mind that the `mongockChangeLog` is a collection that is designed for Mongock to be able to support the process. It's not intended for the user. To check the migration state(as well as repair and any operation that directly target the database), we provide a CLI with a bunch of operations(and much more we are working on). More information [here](https://docs.mongock.io/v5/CLI/index.html)

Said that, yeah you are right. `@Execution` is mandatory, while `@BeforeExecution` is optional, if you want to add an operation before the actual execution and wrap both in the same change unit.


However, if your entire changeUnit is just a non-transactional operation, you can mark the flag `transactional` to false in the `@ChangeUnit` annotation.

**Reply by nazeem-soeltan** [reply]
Hi,

Thanks. Regarding the first point, you're totally right. 

Regarding the `transactional` flag. I missed that one. Thanks for pointing it out. 

Btw, thanks for the fast responses.
